testsuite/classlibrary/TestUGen_RTAlloc.sc

```diff
@@ -70,210 +96,154 @@ TestUGen_RTAlloc : UnitTest {
+		this.assertAllocFail("GrainIn", { GrainIn.ar(1, 1, in:DC.ar(1), maxGrains: 1e5) });
+		this.assertAllocFail("GrainSin", { GrainSin.ar(1, 1, maxGrains: 1e5) });
+
+		this.assertAllocFail("Gendy1", { Gendy1.ar(initCPs: 2**20) });
+		this.assertAllocFail("Gendy2", { Gendy2.ar(initCPs: 2**20) });
+		this.assertAllocFail("Gendy3", { Gendy3.ar(initCPs: 2**20) });
+
+		// these UGens allocate same size as FFT buf
+		// allocating memSizeFloats * 0.75 assures alloc fails in UGen and not in LocalBuf
+		this.assertAllocFail("IFFT", { IFFT(LocalBuf(memSizeFloats * 0.75)) });
+		this.assertAllocFail("SpecPcile", { K2A.ar(SpecPcile.kr(LocalBuf(memSizeFloats * 0.75))) });
+
+		// GVerb needs more tests, as it can fail multiple ways
+		this.assertAllocFail("GVerb", { GVerb.ar(DC.ar(1), 1, maxroomsize: 10000) });
+
+		// PartConv doesn't need a valid buffer when it fails allocation
+		this.assertAllocFail("PartConv", { PartConv.ar(DC.ar(1), 2**20, -1) });
 	}
 
 	// special test for LocalBuf alloc
 	test_LocalBuf {
-		var oneFrame = server.sampleRate.reciprocal;
-		var completed = nil, cond = CondVar();
-		{ K2A.ar(LocalBuf(512)) }.loadToFloatArray(oneFrame, server) { |out|
-			this.assert(out[0] > -1, "should allocate a valid LocalBuf");
-			completed = true;
-			cond.signalOne;
-		};
-		cond.waitFor(1) { completed.notNil };
-		if (completed.isNil) {
+		var out = this.awaitSynthOutput({ K2A.ar(LocalBuf(512)) }, 1);
+		if (out.isNil) {
 			this.assert(false, "successful alloc test should complete");
+		} {
+			var bufnum = out[0];
+			this.assert(bufnum > -1, "should allocate a valid LocalBuf (bufnum: %)".format(bufnum));
 		};
-		completed = nil;
-		{ K2A.ar(LocalBuf(2**20)) }.loadToFloatArray(oneFrame, server) { |out|
-			this.assertFloatEquals(out[0], -1, "should return -1 on alloc fail");
-			completed = true;
-			cond.signalOne;
-		};
-		cond.waitFor(1) { completed.notNil };
-		if (completed.isNil) {
+
+		out = this.awaitSynthOutput({ K2A.ar(LocalBuf(2**20)) }, 1);
+		if (out.isNil) {
 			this.assert(false, "failed alloc test should complete");
+		} {
+			this.assertEquals(out[0], -1, "should return -1 on alloc fail");
 		};
 	}
 
-	// FFT UGens
-	// Test FFT and PV UGens using RTAlloc
-
-	// FFT should return always -1 when alloc fails
+	// FFT UGens: test RTAlloc success and fail
 	test_FFT {
-		var completed = nil, cond = CondVar();
-		var runtime = 2**9 / server.sampleRate / 4;
-		{ FFT(LocalBuf(2**9), DC.ar(1), 0.25) }.loadToFloatArray(runtime, server) { |out|
-			this.assert(out.any(_ > 0), "should return a valid bufnum");
-			completed = true;
-			cond.signalOne;
-		};
-		cond.waitFor(runtime + 1) { completed.notNil };
-		if (completed.isNil) {
+		var out = this.awaitSynthOutput({ FFT(LocalBuf(2**8), DC.ar(1), 0.25) }, 2**6);
+		if (out.isNil) {
 			this.assert(false, "successful alloc test should complete");
+		} {
+			this.assert(out.any(_ > 0), "should return a valid bufnum (bufnum: %)".format(out));
```

Comment:
why is `out` related to the bufnum here? isn't it testing if any of the output samples are `>0`?

Reply:
Yeah, maybe a bit obscure, but FFT normally outputs  a stream of -1. When FFT has processed a frame, which typically takes more than a blockSize, on that block it will output the bufnum instead of -1. It's a way for PV_UGens to know that a new spectral frame is ready and at the same time get the bufnum.
So in this case, out is a stream of -1 and a bufnum > 0. That's why `out.any(_>0)`

Reply:
Ah, thank you! There is always stuff to learn and explore in SC.
Maybe we can add your comment as a comment in a code?

```suggestion
			// FFT normally outputs a stream of -1. When FFT has processed a frame, that block
			// will output the bufnum instead of -1.
			this.assert(out.any(_ > 0), "should return a valid bufnum (bufnum: %)".format(out));
```